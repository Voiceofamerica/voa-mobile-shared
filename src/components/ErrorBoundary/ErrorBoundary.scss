
@import "'~@voiceofamerica/voa-shared/commonStyles/index";

.errorBoundary {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 100vh;
  background: white;

  .content {
    display: flex;
    font-size: 4em;
    font-weight: bold;
    justify-content: center;
  }
  
  .retryButton {
    @include touchable();

    font-size: 2.5em;
    background: $light-grey;
    border-radius: 5px;
    padding: 5px;
  }
}
