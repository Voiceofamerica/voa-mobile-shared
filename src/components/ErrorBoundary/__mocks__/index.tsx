
import * as React from 'react'

export default (props) => (
  React.createElement('error-boundary', { props }, props.children)
)
