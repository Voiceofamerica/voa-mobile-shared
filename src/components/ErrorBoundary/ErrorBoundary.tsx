import * as React from 'react'

import { errorBoundary, content, retryButton } from './ErrorBoundary.scss'

export interface Props {
  error: string | JSX.Element
  retry: string | JSX.Element
  debug?: boolean
}

export interface State {
  hasError: boolean
}

export default class ErrorBoundary extends React.PureComponent<Props, State> {
  state: State = {
    hasError: false,
  }

  componentDidCatch (error) {
    // Display fallback UI
    this.setState({ hasError: true })
    if (this.props.debug) {
      console.error(error)
    }
  }

  onRetry = () => this.setState({ hasError: false })

  render () {
    const { children } = this.props

    if (this.state.hasError) {
      return (
        <div className={errorBoundary}>
          <div className={content}>
            {this.props.error}
          </div>
          <div className={retryButton} onClick={this.onRetry}>
            {this.props.retry}
          </div>
        </div>
      )
    }

    return children as any
  }
}
