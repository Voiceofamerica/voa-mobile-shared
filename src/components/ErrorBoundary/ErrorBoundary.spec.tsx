
import * as React from 'react'
import { create } from 'react-test-renderer'
import { shallow } from 'enzyme'

import ErrorBoundary, { Props, State } from './ErrorBoundary'
import { retryButton } from './ErrorBoundary.scss'

const ErrorComponent = () => { throw new Error() }

describe('<ErrorBoundary />', () => {
  describe('shapshots', () => {
    it('should render only children if no error', () => {
      const element = create((
        <ErrorBoundary error='An Error Occurred' retry='Retry'>
          This should show up
        </ErrorBoundary>
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with an error', () => {
      const element = create((
        <ErrorBoundary error='An Error Occurred' retry='Retry'>
          <ErrorComponent />
        </ErrorBoundary>
      ))

      expect(element).toMatchSnapshot()
    })
  })

  describe('pressing retry', () => {
    it('should reset the state to before an error occurred', () => {
      const wrapper = shallow<Props, State>((
        <ErrorBoundary error='An Error Occurred' retry='Retry' />
      ))
      wrapper.setState({ hasError: true })
      wrapper.find(`.${retryButton}`).simulate('click')

      expect(wrapper.state().hasError).toBe(false)
    })
  })
})
