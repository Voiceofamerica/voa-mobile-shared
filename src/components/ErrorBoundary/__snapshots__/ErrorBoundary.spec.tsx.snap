// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorBoundary /> shapshots should render only children if no error 1`] = `"This should show up"`;

exports[`<ErrorBoundary /> shapshots should render with an error 1`] = `
<div
  className="errorBoundary"
>
  <div
    className="content"
  >
    An Error Occurred
  </div>
  <div
    className="retryButton"
    onClick={[Function]}
  >
    Retry
  </div>
</div>
`;
